File: app/src/main/java/com/mrhuang/demo/chapter2/AudioTrackTask.java
package com.mrhuang.demo.chapter2;

import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.os.Build;

import com.mrhuang.demo.BaseTask;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

public class AudioTrackTask extends BaseTask<Void, Long, String> {

    private String filePath;

    private int rateInHz = 44100;

    int channel = AudioFormat.CHANNEL_OUT_MONO;
    int audioEncoding = AudioFormat.ENCODING_PCM_16BIT;

    private AudioTrack audioTrack;

    public AudioTrackTask() {
    }

    public void play(String filePath) {
        this.filePath = filePath;
        execute();
    }

    public boolean isPlaying() {
        return audioTrack.getPlayState() == AudioTrack.PLAYSTATE_PLAYING;
    }

    public void stop() {
        audioTrack.stop();
    }

    @Override
    protected String doInBackground(Void... voids) {
        File file = new File(filePath);
        if (file.exists()) {
            try {
                int minBufferSize = AudioTrack.getMinBufferSize(rateInHz, channel, audioEncoding);
                byte[] bytes = new byte[minBufferSize];

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    audioTrack = new AudioTrack.Builder()
                            .setAudioFormat(new AudioFormat.Builder()
                                    .setEncoding(audioEncoding)
                                    .setSampleRate(rateInHz)
                                    .setChannelMask(channel)
                                    .build())
                            .setBufferSizeInBytes(minBufferSize)
                            .build();
                } else {
                    audioTrack = new AudioTrack(AudioManager.STREAM_MUSIC, rateInHz, channel,
                            audioEncoding, minBufferSize, AudioTrack.MODE_STREAM);
                }

                audioTrack.play();

                InputStream is = new FileInputStream(filePath);

                int read;
                long current = 0;
                while ((read = is.read(bytes)) > 0) {
                    current += read;
                    audioTrack.write(bytes, 0, bytes.length);
                    publishProgress(current);
                }

                is.close();
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }

        } else {

        }
        return null;
    }
}

File: app/src/main/java/com/mrhuang/demo/chapter2/AudioTrackTask.kt
package com.mrhuang.demo.chapter2

import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.os.Build
import com.mrhuang.demo.BaseTask
import java.io.*

class AudioTrackTask : BaseTask<Void?, Long?, String?>() {
    private var filePath: String? = null
    private val rateInHz = 44100
    var channel = AudioFormat.CHANNEL_OUT_MONO
    var audioEncoding = AudioFormat.ENCODING_PCM_16BIT
    private var audioTrack: AudioTrack? = null
    fun play(filePath: String?) {
        this.filePath = filePath
        execute()
    }

    val isPlaying: Boolean
        get() = audioTrack!!.playState == AudioTrack.PLAYSTATE_PLAYING

    fun stop() {
        audioTrack!!.stop()
    }

    override fun doInBackground(vararg params: Void?): String? {
        val file = File(filePath)
        if (file.exists()) {
            try {
                val minBufferSize = AudioTrack.getMinBufferSize(rateInHz, channel, audioEncoding)
                val bytes = ByteArray(minBufferSize)
                audioTrack = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    AudioTrack.Builder()
                            .setAudioFormat(AudioFormat.Builder()
                                    .setEncoding(audioEncoding)
                                    .setSampleRate(rateInHz)
                                    .setChannelMask(channel)
                                    .build())
                            .setBufferSizeInBytes(minBufferSize)
                            .build()
                } else {
                    AudioTrack(AudioManager.STREAM_MUSIC, rateInHz, channel,
                            audioEncoding, minBufferSize, AudioTrack.MODE_STREAM)
                }
                audioTrack!!.play()
                val `is`: InputStream = FileInputStream(filePath)
                var read: Int
                var current: Long = 0
                while (`is`.read(bytes).also { read = it } > 0) {
                    current += read.toLong()
                    audioTrack!!.write(bytes, 0, bytes.size)
                    publishProgress(current)
                }
                `is`.close()
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        } else {
        }
        return null
    }
}